hare3/gradecast.go

```diff
@@ -0,0 +1,126 @@
+package hare3
+
+import "github.com/spacemeshos/go-spacemesh/common/types"
+
+// GradecastedSet encapsulates the output from gradecast for a given identity.
+type GradecastedSet struct {
+	id     types.Hash20
+	values []types.Hash20
+	grade  uint8
+}
+
+// Gradecaster acts as a specialized value store, it ingests messages that can
+// be retrieved after 3 rounds. Retrieved messages will be graded either 1 or 2
+// based on the input grade and when they were ingested relative to the
+// msgRound. Messages received too late or with insufficient grades will not be
+// retrievable. Subsequent malicious messages can also render previously good
+// messages un-retrievable.
+type Gradecaster interface {
```

Comment:
what is the point of it being an interface? does it simplify testing? i would rather not mock this stuff

Reply:
Initially I started with an interface because I didn't have the implementation, I just added an implementation yesterday, and was considering whether the interface was still useful. I was planning to leave the interfaces till I'd finished all the testing in case a situation came up where we needed to mock something. I can't see that we would need to mock anything but it's no more work to remove the interfaces later rather than now.